include/unifex/async_scope.hpp

```diff
@@ -167,20 +167,44 @@ struct async_scope {
       just_from((Fun&&) fun));
   }
 
+  [[nodiscard]] auto complete() noexcept {
+    return sequence(
+        just_from([this] () noexcept {
+          end_of_scope();
+        }),
+        then(evt_.async_wait(), [this]() noexcept {
+            // make sure to synchronize with all the fetch_subs being done while
+            // operations complete
+            (void)opState_.load(std::memory_order_acquire);
+        }));
+  }
+
   [[nodiscard]] auto cleanup() noexcept {
     return sequence(
         just_from([this]() noexcept {
           request_stop();
         }),
-        then(evt_.async_wait(), [this]() noexcept {
-          // make sure to synchronize with all the fetch_subs being done while
-          // operations complete
-          (void)opState_.load(std::memory_order_acquire);
-        }));
+        complete());
+  }
+
+  inplace_stop_token get_stop_token() noexcept {
+    return stopSource_.get_token();
+  }
+
+  void request_stop() noexcept {
+    stopSource_.request_stop();
   }
 
  private:
+  struct stop_callback {
+    async_scope& self_;
+    void operator()() noexcept {
+      self_.end_of_scope();
+    }
+  };
+
   inplace_stop_source stopSource_;
+  inplace_stop_token::template callback_type<stop_callback> stopCallback_;
```

Comment:
It feels unfortunate that this diff adds a stop callback, especially given that the stop source is encapsulated behind the interface and so the type controls when stop is requested.

Perhaps `end_of_scope` should take a `bool` indicating whether to request stop, and `request_stop` and `cleanup` can call it with `true`, while `complete` can call it with `false`.

Reply:
I originally planned to add a get_stop_source() method and that was why I took this approach. Ultimately, I left that method out because I was unsure of exporting a reference to the inplace_stop_source. In the current state of this PR your suggestion will work - but I hate bool params - they are always undocumented when reading the source.

Reply:
BTW, What is unfortunate about the callback? the overhead of calling spawn once should make this cost disappear, right?

Reply:
> What is unfortunate about the callback?

I think the thing that tingles my Spidey sense is that the callback adds "spooky action at a distance" between `request_stop` and `end_of_scope`.  It's both lexically disconnected and a more complicated coupling.  Incidentally, in the current implementation, `cleanup` deterministically calls `end_of_scope` twice (once in the stop callback and once in `complete`); it's idempotent, but it's also an atomic read-modify-write plus a branch.

I suspect you're right that the overhead of the complexity is small, especially compared to other operations that this type performs, but it feels like unnecessary overhead that's _also_ more complicated.

> I hate bool params - they are always undocumented when reading the source.

Fair; I also find `bool` params unwieldy.  My fall-back suggestion instead of `bool`s is usually to think of `bool` as the most generic possible `enum` and to use a more specific one instead, e.g.
```
enum class EndScopeStrategy {
  RequestStop,
  DoNotRequestStop
};
```

Another option would be to pass the address of the stop source and only request stop if it's not `nullptr`.

Last, I could imagine factoring it thus:
```
void stop_and_end_scope() {
  stopSource_.request_stop();
  end_scope();
}

void end_scope() {
  // ... muck about with opState_
}
```

I don't prefer this last option because it leaves a window where someone trying to spawn a task can observe "it's still ok to spawn work" _and_ "that work should immediately stop", but this is a) a narrow window, and b) probably not a real problem.